### ISSUE TITLE
lastColumLetter Bug in Xlsx::listWorksheetInfo($pFilename)

### ISSUE THREAD

**Issue by danzy83** [PHPOffice/PhpSpreadsheet#470]
This is:

 ``` - a bug report```

I have noticed a bug in ```Xlsx::listWorksheetInfo($pFilename)``` which I am reporting here for a fix.

### The Bug Issue

When ```Xlsx::listWorksheetInfo($pFilename)``` is called, it digs for file information such as totalRows, totalColumns, lastColumnIndex, and lastColumnLetter. However, ```lastColumnIndex``` and ```lastColumnLetter``` are wrongly reported when used columns does not start from the first column.

### What is the current behavior?

In the following attached file, the first column is unused but the only text for testing (Hello) is entered in cell B of row 1 (that is, ```B1```)

[test.xlsx](https://github.com/PHPOffice/PhpSpreadsheet/files/1940519/test.xlsx)

In this case calling ```Xlsx::listWorksheetInfo($pFilename)``` returns ```A``` as the lastColumnLetter instead of cell ```B```. Upon navigating the ```Xlsx.php``` file, I noticed the problem emanates from how ```lastColumnIndex``` is determined from the source:

```php
<?php
    // Some code above
    // ......

    // lastColumnIndex is determined below
   $tmpInfo['lastColumnIndex'] = $tmpInfo['totalColumns'] - 1;
   $tmpInfo['lastColumnLetter'] = Coordinate::stringFromColumnIndex($tmpInfo['lastColumnIndex'] + 1);
?>
```

Obviosuly, the problem is in the use of ```totalColumns``` to determine ```lastColumnIndex```. It appears ```totalColumns``` ignores initial unused columns  and returns the number of used columns. This makes

      $tmpInfo['lastColumnIndex'] = $tmpInfo['totalColumns'] - 1; 

evaluate to ```0``` making ```lastColumnLetter``` to be ```A```.

Also, ```lastColumnLetter``` deviation is also dependent on the number of preceding unused columns. For example, if columns A and B are unused and the text entered in column C, lastColumnLetter is still reported as ```A``` instead of column letter ```C```. This is because ```totalColumns``` is still reported as 1 making lastColumnIndex to be ```0``` and hence lastColumnLetter ```A```.

### Which versions of PhpSpreadsheet and PHP are affected?

This problem dates back to the era of PHPExcel and still found in the current version of PhpSpreadsheet.

**Comment by patrickvuarnoz** [comment]
I had the same issue as reported by danzy83. The problem seems to be that the current implementation just counts `<c>` (column) tags for each `<row>` in the worksheet XML and reports the maximum count as total columns. 

It seems though that in XLSX only columns containing values make it into the the XML and create a `<c>` tag. So, as soon as you have a full column with no value in it the reported total columns will be wrong. And with that the last column index.

The current code looks like this (version 1.3.1, PhpSpreadsheet/Reader/Xlsx.php, starting at line 222):

```
$currCells = 0;
while ($xml->read()) {
    if ($xml->name == 'row' && $xml->nodeType == XMLReader::ELEMENT) {
        $row = $xml->getAttribute('r');
        $tmpInfo['totalRows'] = $row;
        $tmpInfo['totalColumns'] = max($tmpInfo['totalColumns'], $currCells);
        $currCells = 0;
    } elseif ($xml->name == 'c' && $xml->nodeType == XMLReader::ELEMENT) {
        ++$currCells;
    }
}
$tmpInfo['totalColumns'] = max($tmpInfo['totalColumns'], $currCells);
```

A fix to the issue might be to not count cells but to actually analyse the cell refs. By that we can also skip to analyse the refs of the rows since the cells carry the row number as well: 

```
while ($xml->read()) {
    if ($xml->name == 'c' && $xml->nodeType == XMLReader::ELEMENT) {
        $ref = $xml->getAttribute('r');
        $coord = Coordinate::coordinateFromString($ref);
        $tmpInfo['totalColumns'] = max($tmpInfo['totalColumns'], Coordinate::columnIndexFromString($coord[0]));
        $tmpInfo['totalRows'] = max($tmpInfo['totalRows'], $coord[1]);
    }
}
```

This solution worked for me.

**Comment by PowerKiKi** [comment]
Would you like to create a PR for this fix ?

**Comment by patrickvuarnoz** [comment]
Unfortunately I'm not completely aware of how create a PR and I don't have the time to create enough test cases to fully verify the proposed fix with all kinds of different XSLX files. The proposed fix worked 100% for our needs in a controlled environment.

Just wanted to share our solution since I stumbled over this bug report searching for one.

If anybody can take the necessary steps to help this fix making it into the main repo, I'd very much appreciate that.

**Comment by sbreiler** [comment]
@PowerKiKi could you/someone re-open this issue?

**Comment by PowerKiKi** [comment]
Your reasoning seems to make sense to me. Unfortunately I don't have answers for your questions. You'll probably have to read XLSX spec and/or experiment with real life files produced by different tools.

**Comment by Phjester** [comment]
I found this problem in version 1.17.1. I installed version 1.16.0 and its worked

**Comment by oleibman** [comment]
Seems to have been corrected in Release 1.19.
